src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java

```diff
@@ -277,33 +188,44 @@ public static boolean checkBuiltInFunctionExists(FunctionIdentifier identifier)
 
     public static RuntimeIterator getBuiltInFunctionIterator(
             FunctionIdentifier identifier,
-            IteratorMetadata metadata,
-            List<RuntimeIterator> arguments
+            List<RuntimeIterator> arguments,
+            IteratorMetadata metadata
     ) {
-        Class<? extends RuntimeIterator> functionClass = builtInFunctions.get(identifier);
-        try {
-            Constructor<? extends RuntimeIterator> ctor = functionClass.getConstructor(
-                List.class,
-                IteratorMetadata.class
+        BuiltinFunction builtinFunction = builtInFunctions.get(identifier);
+        for (int i = 0; i < arguments.size(); i++) {
+            arguments.set(
+                i,
+                new TypePromotionIterator(
+                        arguments.get(i),
+                        builtinFunction.getSignature().getParameters().get(i),
+                        metadata
+                )
             );
-            return ctor.newInstance(arguments, metadata);
-        } catch (Exception e) {
-            throw new RuntimeException(e.getMessage());
         }
+        return new CheckReturnTypeIterator(
+                new TypePromotionIterator(
+                        new BuiltinFunctionCallIterator(builtinFunction, arguments, metadata),
```

Comment:
I am not sure I understand why this new wrapping object is created, rather than simply sticking to

ctor.newInstance(arguments, metadata)

Could you explain what your reasoning is, or if you ran into any troubles just using the iterator instance as is?

This creates a runtime overhead with additional object constructions, which is costly. For builtin functions, the iterator class is "ready to use off the shelf" without extra machinery. This makes them faster as, again, they may be called billions of times.

Normally, there is no need to modify anything in the function calls iterators themselves -- this is outside of the scope of your thesis. This is the point of having this method as the point of entry to simply wrap iterators inside the treat as and type promotion operators.

Reply:
I introduced that iterator for a couple of reasons:

- I thought that merging builtin functions and user-defined functions was a plan for the future, so I tried to make them already as similar as possible in the way they are used as well, i.e. introducing this point-of-entry for builtin functions.

- To have a single place where I could catch the TreatException and throw a corresponding (i.e. UnexpectedTypeException with message _"invalid argument type"_). That is the same reason for CheckReturnTypeIterator to throw UnexpectedTypeException with message _"invalid return type"_. If we don't want that I guess we will have to say something more general like "string cannot be promoted to type double" without saying that this error comes from the argument or from the return type (if we also want to get rid of CheckReturnTypeIterator).

Reply:
Thank you for your answer and these details. This is very useful.

> I introduced that iterator for a couple of reasons:
> 
> * I thought that merging builtin functions and user-defined functions was a plan for the future, so I tried to make them already as similar as possible in the way they are used as well, i.e. introducing this point-of-entry for builtin functions.

The two APIs have already been unified, in the sense that both kinds of iterators are created from methods in the function catalogue that both pass a list of runtime iterator children, and return a runtimeiterator.

This methods are called getBuiltInFunctionIterator and buildUserDefinedFunctionCallIterator.

Doing the type checks in these two methods (in the same way) will take care of absolutely everything -- even dynamic function calls. The change in buildUserDefinedFunctionCallIterator, however, can be done in a later PR once this one is finished, but I anticipate it should only be a few lines of code.

There is no plan to merge the iterators beyond that, because of performance: builtin functions must stay streamlined in the way they are as they will be called billions of times. Any extra wrapping machinery would slow down the system.

Thank you for your kind understanding that function call iterators are out of scope in the context of type checking, and should be left untouched. It is important that the other thesis remains unimpacted by these changes, and that everybody can focus on their topic.

> * To have a single place where I could catch the TreatException and throw a corresponding (i.e. UnexpectedTypeException with message _"invalid argument type"_). That is the same reason for CheckReturnTypeIterator to throw UnexpectedTypeException with message _"invalid return type"_. If we don't want that I guess we will have to say something more general like "string cannot be promoted to type double" without saying that this error comes from the argument or from the return type (if we also want to get rid of CheckReturnTypeIterator).

I understand, thank you for this explanation. The design with catching and throwing again has an issue: in getRDDAux, no exception would be thrown as this is a lazy evaluation: the exception would only pop up at the next action in the pipeline.

So, any information that must appear in the error message could/should also be passed as an argument to the treat as iterator, which can directly throw the appropriate error codes early on without catching them later in the pipeline.

Indeed, CheckReturnTypeIterator would then also be superfluous, which also simplifies the code and improves performance (less method calls, less exception catching). This should also considerably diminish the size of this PR (which is the goal of these entry points, which isolate type checking from the function call machinery).

Reply:
> There is no plan to merge the iterators beyond that, because of performance: builtin functions must stay streamlined in the way they are as they will be called billions of times. Any extra wrapping machinery would slow down the system.

Ok, got it! I thought the final plan was to have a single iterator that would be the entry point for both, but it actually would lose in performance to add another layer of iterator as you rightfully pointed out. 
I removed BuiltinFunctionCallIterator and CheckReturnTypeIterator so, now, TypePromotionIterator is parametrized with the exception message that should be thrown depending on whether this iterator is wrapping parameters or the function iterator (so checking the return type). 


> The design with catching, and throwing again a different exception, has one issue: in getRDDAux, no exception would be thrown as this is a lazy evaluation: the exception would only pop up at the next action in the pipeline.
So, any information that must appear in the error message could/should also be passed as an argument to the treat as iterator, which can directly throw the appropriate error codes early on without catching them later in the pipeline.

I parametrized TreatAsIterator to either throw TreatException or UnexpectedTypeException (as discussed in #379 as well). Although TypePromotionException will have to go through the try-catch of the TreatException because it will have to check for promotion before forwarding the exception further into the pipeline. Do you think the problem you mentioned because of lazy evaluation would still persist? If so, does that mean that we cannot check for promotion with TypePromotionIterator if we are not in local evaluation? I would appreciate if you could point out to me any error (if you see any) in the getRDDAux() function and of TypePromotionIterator, and its closure. Thank you

Reply:
> Do you think the problem you mentioned because of lazy evaluation would still persist? If so, does that mean that we cannot check for promotion with TypePromotionIterator if we are not in local evaluation? I would appreciate if you could point out to me any error (if you see any) in the getRDDAux() function and of TypePromotionIterator, and its closure. Thank you

I think that my comments should address this question, as this is closely related. The implementation of getRDDAux will then follow very naturally.

Reply:
> Ok, got it! I thought the final plan was to have a single iterator that would be the entry point for both, but it actually would lose in performance to add another layer of iterator as you rightfully pointed out.
> I removed BuiltinFunctionCallIterator and CheckReturnTypeIterator so, now, 

Very nice. Thanks a lot for taking care of this.